### PR TITLE
Add helm chart for pattern 3

### PR DESCRIPTION
**Pull request by Chathuru**
Add helm chart for pattern 3

### CODE REVIEW

File: advanced/am-pattern-2/README.md [hunk 1]
```diff
@@ -9,10 +9,33 @@
```
Comment (chirangaalwis) [inline advanced/am-pattern-2/README.md:9]: @Chathuru Please recheck the README instructions. I saw missing sub headings in the case of instructions for running this chart from the Helm Hub (e.g. see [sample](https://github.com/wso2/kubernetes-is/tree/master/advanced/is-pattern-1#install-chart-from-wso2-helm-chart-repository)).

Also, please update the Helm chart version to map with the new release.

File: advanced/am-pattern-2/templates/am-analytics/worker-2/wso2am-pattern-2-am-analytics-worker-conf.yaml [hunk 2]
```diff
@@ -28,7 +28,7 @@ data:
   {{- else }}
```
Comment (chirangaalwis) [inline advanced/am-pattern-2/templates/am-analytics/worker-2/wso2am-pattern-2-am-analytics-worker-conf.yaml:28]: @Chathuru let us rename the ConfigMap as highlighted earlier.

File: advanced/am-pattern-1/templates/am/wso2am-pattern-1-am-volume-claim.yaml [hunk 3]
```diff
@@ -23,14 +23,7 @@ spec:
   resources:
```
Comment (chirangaalwis) [inline advanced/am-pattern-1/templates/am/wso2am-pattern-1-am-volume-claim.yaml:23]: @Chathuru Let us rename this files to `wso2am-pattern-1-am-volume-claims.yaml` as it contains multiple PVCs.

File: advanced/am-pattern-2/templates/am/pub-store-tm/instance-1/wso2am-pattern-2-instance-1-am-volume-claim.yaml [hunk 4]
```diff
@@ -23,14 +23,7 @@ spec:
   resources:
```
Comment (chirangaalwis) [inline advanced/am-pattern-2/templates/am/pub-store-tm/instance-1/wso2am-pattern-2-instance-1-am-volume-claim.yaml:23]: @Chathuru Let us rename this file as it contains multiple PVCs.

File: advanced/am-pattern-2/templates/am/pub-store-tm/instance-2/wso2am-pattern-2-instance-1-am-volume-claim.yaml [hunk 5]
```diff
@@ -23,14 +23,7 @@ spec:
   resources:
```
Comment (chirangaalwis) [inline advanced/am-pattern-2/templates/am/pub-store-tm/instance-2/wso2am-pattern-2-instance-1-am-volume-claim.yaml:23]: @Chathuru Let us rename this file as it contains multiple PVCs. Also, the filename refers to `instance-1`.

File: advanced/am-pattern-3/requirements.yaml [hunk 6]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 dependencies:
   - name: mysql-am
-    version: "3.1.0-2"
+    version: "3.1.0-1"
```
Comment (chirangaalwis) [inline advanced/am-pattern-3/requirements.yaml:16]: @Chathuru the upcoming release is `3.1.0-2`. So, no need to change.

File: advanced/am-pattern-3/templates/am/tm/instance-1/wso2am-pattern-3-am-tm-1-deployment.yaml [hunk 7]
```diff
@@ -12,44 +12,20 @@
...
-
 apiVersion: apps/v1
-kind: StatefulSet
+kind: Deployment
 metadata:
   name: wso2am-pattern-3-am-tm-1-deployment
   namespace: {{ .Release.Namespace }}
 spec:
+  replicas: {{ .Values.wso2.deployment.am.tm.replicas }}
+  minReadySeconds: {{ .Values.wso2.deployment.am.minReadySeconds }}
```
Comment (chirangaalwis) [inline advanced/am-pattern-3/templates/am/tm/instance-1/wso2am-pattern-3-am-tm-1-deployment.yaml:22]: @Chathuru @Aaquiff @ThilinaManamgoda Do we require `minReadySeconds`? I suppose we can consider the Pod to be ready as soon as the probes are completed (see [issue](https://stackoverflow.com/questions/53239081/how-does-minreadyseconds-affect-readiness-probe)).

File: CHANGELOG.md [hunk 8]
```diff
@@ -7,3 +7,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Introduce Helm chart for WSO2 APIM version 3.1.0 deployment pattern 1
 - Introduce Helm chart for WSO2 APIM version 3.1.0 deployment pattern 2
 - Introduce Simplified setup for WSO2 APIM version 3.1.0
+
+## [v3.1.0.2]
+- Introduce Helm chart for WSO2 APIM version 3.1.0 deployment pattern 3
+- Add support for Custom Kubernetes Storage Class
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply CHANGELOG.md:8]: Edit the change log in this commit https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5

File: CHANGELOG.md [hunk 9]
```diff
@@ -7,3 +7,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Introduce Helm chart for WSO2 APIM version 3.1.0 deployment pattern 1
 - Introduce Helm chart for WSO2 APIM version 3.1.0 deployment pattern 2
 - Introduce Simplified setup for WSO2 APIM version 3.1.0
+
+## [v3.1.0.2]
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply CHANGELOG.md]: Add release date to changelog in https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5 commit

File: advanced/am-pattern-1/README.md [hunk 10]
```diff
@@ -160,9 +160,7 @@ If you do not have an active WSO2 subscription, do not change the parameters `ws
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply advanced/am-pattern-1/README.md:160]: Update the readme command in https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5

File: advanced/am-pattern-1/templates/am-analytics/worker-2/wso2am-pattern-1-am-analytics-worker-conf.yaml [hunk 11]
```diff
@@ -28,7 +28,7 @@ data:
   {{- else }}
 
   {{- $file := .Files }}
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply advanced/am-pattern-1/templates/am-analytics/worker-2/wso2am-pattern-1-am-analytics-worker-conf.yaml]: Fixed https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5

File: advanced/am-pattern-1/templates/am/instance-2/wso2am-pattern-1-instance-2-am-volume-claim.yaml [hunk 12]
```diff
@@ -23,14 +23,7 @@ spec:
...
       storage: 2Gi
-  {{- if and (.Values.wso2.deployment.persistentRuntimeArtifacts.cloudProvider) (eq .Values.wso2.deployment.persistentRuntimeArtifacts.cloudProvider "internal-nfs") }}
-  storageClassName: nfs
-  {{- else if and (.Values.wso2.deployment.persistentRuntimeArtifacts.cloudProvider) (eq .Values.wso2.deployment.persistentRuntimeArtifacts.cloudProvider "external-nfs") }}
-  storageClassName: ""
-  selector:
-    matchLabels:
-      purpose: wso2am-pattern-1-instance-2-carbon-db
-  {{- end }}
+  storageClassName: {{ .Values.wso2.deployment.persistentRuntimeArtifacts.storageClass }}
```
[the remark this answers is not in the document]
Comment (Chathuru) [reply advanced/am-pattern-1/templates/am/instance-2/wso2am-pattern-1-instance-2-am-volume-claim.yaml:26]: In pattern 1 and 2, I only change to use custom storage classes and remove duplicate analytic worker configs. Also, Change the chart versions and requirement.yaml. Other than that I haven’t try to change the am's statefulset to deployments as I didn’t have enough time.

[the remark this answers is not in the document]
Comment (Chathuru) [reply advanced/am-pattern-1/templates/am/instance-2/wso2am-pattern-1-instance-2-am-volume-claim.yaml:26]: Also, I could able to run 2 TM instances without any error logs in pattern 3

File: advanced/am-pattern-2/README.md (hunk 1 above)
Comment (Chathuru) [reply advanced/am-pattern-2/README.md:9]: Update the readme to the correct helm chart version. What are the missing sub headings

File: advanced/am-pattern-2/templates/am-analytics/worker-2/wso2am-pattern-2-am-analytics-worker-conf.yaml (hunk 2 above)
Comment (Chathuru) [reply advanced/am-pattern-2/templates/am-analytics/worker-2/wso2am-pattern-2-am-analytics-worker-conf.yaml:28]: Fix in https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5

File: advanced/am-pattern-2/templates/am/pub-store-tm/instance-1/wso2am-pattern-2-instance-1-am-volume-claim.yaml (hunk 4 above)
Comment (Chathuru) [reply advanced/am-pattern-2/templates/am/pub-store-tm/instance-1/wso2am-pattern-2-instance-1-am-volume-claim.yaml:23]: Renamed the file in https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5

File: advanced/am-pattern-1/templates/am/wso2am-pattern-1-am-volume-claim.yaml (hunk 3 above)
Comment (Chathuru) [reply advanced/am-pattern-1/templates/am/wso2am-pattern-1-am-volume-claim.yaml:23]: Renamed the file in https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5

File: advanced/am-pattern-2/templates/am/pub-store-tm/instance-2/wso2am-pattern-2-instance-1-am-volume-claim.yaml (hunk 5 above)
Comment (Chathuru) [reply advanced/am-pattern-2/templates/am/pub-store-tm/instance-2/wso2am-pattern-2-instance-1-am-volume-claim.yaml:23]: Renamed https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5

File: advanced/am-pattern-3/requirements.yaml (hunk 6 above)
Comment (Chathuru) [reply advanced/am-pattern-3/requirements.yaml]: Update to the new dependency version in all 3 patterns in https://github.com/wso2/kubernetes-apim/pull/365/commits/656646538e8cc1b8d8fbc51cb3d206a1c58101e5

File: advanced/am-pattern-2/README.md (hunk 1 above)
Comment (chirangaalwis) [reply advanced/am-pattern-2/README.md:9]: @Chathuru the one I found is `Install Chart From WSO2 Helm Chart Repository`. Please recheck the whole doc again.

File: advanced/am-pattern-2/templates/am/pub-store-tm/instance-2/wso2am-pattern-2-instance-1-am-volume-claims.yaml [hunk 13]
```diff
@@ -23,14 +23,7 @@ spec:
   resources:
```
Comment (chirangaalwis) [inline advanced/am-pattern-2/templates/am/pub-store-tm/instance-2/wso2am-pattern-2-instance-1-am-volume-claims.yaml:23]: @Chathuru the filename should change to `wso2am-pattern-2-instance-2-am-volume-claims.yaml`.

File: advanced/am-pattern-3/templates/am/gateway/wso2am-pattern-3-am-gateway-deployment.yaml [hunk 14]
```diff
@@ -128,3 +131,7 @@ spec:
       - name: wso2am-pattern-3-am-gateway-conf
         configMap:
           name: wso2am-pattern-3-am-gateway-conf
+      - name: wso2am-pattern-3-conf-entrypoint
```
Comment (chirangaalwis) [inline advanced/am-pattern-3/templates/am/gateway/wso2am-pattern-3-am-gateway-deployment.yaml:134]: @Chathuru why do we need to override the entry point here?

File: advanced/am-pattern-3/templates/am/km/wso2am-pattern-2-am-km-deployment.yaml [hunk 15]
```diff
@@ -99,6 +99,9 @@ spec:
         - name: wso2am-pattern-3-am-km-conf
```
Comment (chirangaalwis) [inline advanced/am-pattern-3/templates/am/km/wso2am-pattern-2-am-km-deployment.yaml:99]: @Chathuru why do we need to override the entry point here?

File: advanced/am-pattern-3/templates/am/tm/instance-2/wso2am-pattern-3-am-tm-2-deployment.yaml [hunk 16]
```diff
@@ -119,6 +95,9 @@ spec:
         - name: wso2am-pattern-3-am-tm-2-conf
           mountPath: /home/wso2carbon/wso2-config-volume/repository/conf/deployment.toml
           subPath: deployment.toml
+        - name: wso2am-pattern-3-conf-entrypoint
```
Comment (chirangaalwis) [inline advanced/am-pattern-3/templates/am/tm/instance-2/wso2am-pattern-3-am-tm-2-deployment.yaml:98]: @Chathuru why do we need to override the entry point here?

File: advanced/am-pattern-3/templates/am/wso2am-pattern-3-conf-entrypoint.yaml [hunk 17]
```diff
@@ -0,0 +1,88 @@
+# Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chirangaalwis) [inline advanced/am-pattern-3/templates/am/wso2am-pattern-3-conf-entrypoint.yaml:1]: @Chathuru I don't think we need to override the entry point script (except in the case of Publisher and DevPortal where solr indexing is used), IIRC. Profile optimization supported within the image itself. If the images are not available in Docker Hub, we need to push them to the Docker Hub.

File: advanced/am-pattern-2/templates/am/pub-store-tm/instance-2/wso2am-pattern-2-instance-1-am-volume-claims.yaml (hunk 13 above)
Comment (Chathuru) [reply advanced/am-pattern-2/templates/am/pub-store-tm/instance-2/wso2am-pattern-2-instance-1-am-volume-claims.yaml]: Fix https://github.com/wso2/kubernetes-apim/pull/365/commits/3588e9f5044c3947741ef019ecda0ee150daad18

File: advanced/am-pattern-3/templates/am/gateway/wso2am-pattern-3-am-gateway-deployment.yaml (hunk 14 above)
Comment (Chathuru) [reply advanced/am-pattern-3/templates/am/gateway/wso2am-pattern-3-am-gateway-deployment.yaml]: fix https://github.com/wso2/kubernetes-apim/pull/365/commits/3588e9f5044c3947741ef019ecda0ee150daad18

File: advanced/am-pattern-3/templates/am/wso2am-pattern-3-conf-entrypoint.yaml (hunk 17 above)
Comment (Chathuru) [reply advanced/am-pattern-3/templates/am/wso2am-pattern-3-conf-entrypoint.yaml]: Fix https://github.com/wso2/kubernetes-apim/pull/365/commits/3588e9f5044c3947741ef019ecda0ee150daad18

File: advanced/am-pattern-3/templates/am/km/wso2am-pattern-2-am-km-deployment.yaml (hunk 15 above)
Comment (Chathuru) [reply advanced/am-pattern-3/templates/am/km/wso2am-pattern-2-am-km-deployment.yaml]: Fix https://github.com/wso2/kubernetes-apim/pull/365/commits/3588e9f5044c3947741ef019ecda0ee150daad18

File: advanced/am-pattern-3/templates/am/tm/instance-2/wso2am-pattern-3-am-tm-2-deployment.yaml (hunk 16 above)
Comment (Chathuru) [reply advanced/am-pattern-3/templates/am/tm/instance-2/wso2am-pattern-3-am-tm-2-deployment.yaml]: Fix https://github.com/wso2/kubernetes-apim/pull/365/commits/3588e9f5044c3947741ef019ecda0ee150daad18

**Pull request merged**